log-parser/cpp/log_parser_hs.cpp

```diff
@@ -0,0 +1,619 @@
+#include "log_parser.h"
+#include <cassert>
+#include <cstdlib>
+#include <cstdio>
+#include <cctype>
+#include <thread>
+#include <memory>
+#include <algorithm>
+#include <iterator>
+#include <fstream>
+#include <sstream>
+#include <ctime>
+
+#if defined(LP_REGEX_HYPERSCAN)
+#include <hs/hs.h>
+#include <hs/hs_compile.h>
+
+class RegexHS {
+	hs_database *db;
+public:
+	RegexHS(std::string_view pattern) {
+		hs_compile_error_t *err = nullptr;
+		hs_compile(pattern.data(), HS_FLAG_SINGLEMATCH || HS_MODE_SOM_HORIZON_SMALL, HS_MODE_BLOCK, nullptr, &db, &err);
+		if (err != nullptr) {
+			perror(err->message);
+			hs_free_compile_error(err);
+		}
+	}
+	hs_database *get_db() {
+		return db;
+	}
+	bool search(std::string_view str, hs_scratch_t *scratch) {
+		return HS_SUCCESS == hs_scan(db, str.data(), str.length(), 0, scratch, nullptr, nullptr);
+	}
+	~RegexHS() {
+		hs_free_database(db);
+	}
+};
+// One scratch per thread (according to doc), and it should be created before the search/scan
+std::vector<hs_scratch_t*> scratch_cache;
```

Comment:
[`thread_local`](https://en.cppreference.com/w/cpp/keyword/thread_local) would maybe also work here if you didn't it existed (but it would not work with that loop copying stuff later on).

Reply:
I didn't know about this. Maybe I'll try using it.